### PR TITLE
chore(java): Disallow writing meta classdef when obj is null

### PR DESCRIPTION
**Pull request by LiangliangSui**
## What does this PR do?

When obj is null, if shareMeta is enabled, no need to write meta classdef

## Does this PR introduce any user-facing change?

- [ ] Does this PR introduce any public API change?
- [ ] Does this PR introduce any binary protocol compatibility change?

### CODE REVIEW

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 1]
```diff
@@ -310,12 +310,13 @@ private void write(MemoryBuffer buffer, Object obj) {
...
     // reduce caller stack
-    if (!refResolver.writeRefOrNull(buffer, obj)) {
+    boolean isNull = refResolver.writeRefOrNull(buffer, obj);
+    if (!isNull) {
       ClassInfo classInfo = classResolver.getOrUpdateClassInfo(obj.getClass());
       classResolver.writeClass(buffer, classInfo);
       writeData(buffer, classInfo, obj);
     }
-    if (shareMeta) {
+    if (shareMeta && !isNull) {
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:319]: The deserialization will still read class meta, but the offset will be -1, then the read will fail. Could you add an UT for such case?

Comment (LiangliangSui) [reply java/fury-core/src/main/java/org/apache/fury/Fury.java:322]: Done.

CI failure is not related to the current changes, fix it in https://github.com/apache/fury/pull/1691.

**Comment by chaokunyang** [comment]
The code looks good to me. But do we really to optimize this? When value is null, the ClassDef list is empty, only one byte are written in such cases

**Comment by LiangliangSui** [comment]
> The code looks good to me. But do we really to optimize this? When value is null, the ClassDef list is empty, only one byte are written in such cases

There will be two more write overheads (putInt32, writeVarUint32Small7), which is also a small problem I found while reading the code. If there is no negative impact, I think it can be merged.

**Comment by LiangliangSui** [comment]
Hi @chaokunyang , could you reply to this PR?

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 2]
```diff
@@ -307,23 +309,26 @@ private void write(MemoryBuffer buffer, Object obj) {
...
       buffer.putInt32(startOffset, buffer.writerIndex());
       classResolver.writeClassDefs(buffer);
     }
   }
 
   private void xserializeInternal(MemoryBuffer buffer, Object obj) {
     int startOffset = buffer.writerIndex();
-    buffer.writeInt32(-1); // preserve 4-byte for nativeObjects start offsets.
+    buffer.writeInt32(
+        Constants.INVALID_OFFSITE); // preserve 4-byte for nativeObjects start offsets.
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:331]: This is an internal, could we keep it inside this class

Comment (LiangliangSui) [reply java/fury-core/src/main/java/org/apache/fury/Fury.java:331]: Also used in ClassResolver https://github.com/apache/fury/pull/1686/files#diff-849c034dbc915d01d33d09221ec8c78b290ac53a845e3197ab648ec05bea480aR1427-R1430

**Comment by chaokunyang** [comment]
How about checking meta list size and remove this part of code from class resolver into Fury

**Comment by LiangliangSui** [comment]
I don't think it's necessary. Reading `classDefOffset` and reading `numClassDefs` should be what `readClassDefs` should do. We shouldn't distribute what `readClassDefs` should do to other functions.

Why would we do this? Just to move `Constants.INVALID_OFFSITE` to `Fury`?

**Comment by chaokunyang** [comment]
> I don't think it's necessary. Reading `classDefOffset` and reading `numClassDefs` should be what `readClassDefs` should do. We shouldn't distribute what `readClassDefs` should do to other functions.
> 
> Why would we do this? Just to move `Constants.INVALID_OFFSITE` to `Fury`?

When object is not null, but the type is internal types such pritimitive type, string or ArrayList/HashMap, the meta list will be empty too

**Comment by LiangliangSui** [comment]
> but the type is internal types such pritimitive type

Not only internal types, but also the meta list of registered classes(`classInfo.classId != NO_CLASS_ID`) is empty.

> How about checking meta list size and remove this part of code from class resolver into Fury

```java
  private void write(MemoryBuffer buffer, Object obj) {
    int startOffset = buffer.writerIndex();
    boolean shareMeta = config.isMetaShareEnabled();
    if (shareMeta) {
      buffer.writeInt32(INVALID_OFFSITE); // preserve 4-byte for nativeObjects start offsets.
    }
    // reduce caller stack
    boolean isNull = refResolver.writeRefOrNull(buffer, obj);
    if (!isNull) {
      ClassInfo classInfo = classResolver.getOrUpdateClassInfo(obj.getClass());
      classResolver.writeClass(buffer, classInfo);
      writeData(buffer, classInfo, obj);
    }
    // **Just determine the meta list size.**
    if (shareMeta && !isNull && !serializationContext.getMetaContext().writingClassDefs.isEmpty()) {
      buffer.putInt32(startOffset, buffer.writerIndex());
      classResolver.writeClassDefs(buffer);
    }
  }
```
Is it possible to determine the meta list size at the end?

**Comment by LiangliangSui** [comment]
> > but the type is internal types such pritimitive type
> 
> Not only internal types, but also the meta list of registered classes(`classInfo.classId != NO_CLASS_ID`) is empty.
> 
> > How about checking meta list size and remove this part of code from class resolver into Fury
> 
> ```java
>   private void write(MemoryBuffer buffer, Object obj) {
>     int startOffset = buffer.writerIndex();
>     boolean shareMeta = config.isMetaShareEnabled();
>     if (shareMeta) {
>       buffer.writeInt32(INVALID_OFFSITE); // preserve 4-byte for nativeObjects start offsets.
>     }
>     // reduce caller stack
>     boolean isNull = refResolver.writeRefOrNull(buffer, obj);
>     if (!isNull) {
>       ClassInfo classInfo = classResolver.getOrUpdateClassInfo(obj.getClass());
>       classResolver.writeClass(buffer, classInfo);
>       writeData(buffer, classInfo, obj);
>     }
>     // **Just determine the meta list size.**
>     if (shareMeta && !isNull && !serializationContext.getMetaContext().writingClassDefs.isEmpty()) {
>       buffer.putInt32(startOffset, buffer.writerIndex());
>       classResolver.writeClassDefs(buffer);
>     }
>   }
> ```
> 
> Is it possible to determine the meta list size at the end?

@chaokunyang Any other suggestions? If so, please let me know.

**Comment by chaokunyang** [comment]
> the meta list of registered classes(`classInfo.classId != NO_CLASS_ID`) is empty.

For the meta list of registered classes(classInfo.classId != NO_CLASS_ID), we still need to write meta, this needs being done before we set scopde meta share as the default compatible mode

**Comment by chaokunyang** [comment]
We can check meta size before writing meta

**Comment by LiangliangSui** [comment]
> > the meta list of registered classes(`classInfo.classId != NO_CLASS_ID`) is empty.
> 
> For the meta list of registered classes(classInfo.classId != NO_CLASS_ID), we still need to write meta, this needs being done before we set scopde meta share as the default compatible mode

I mean if the current class has been registered, the information of the current class will not be written back to meta.

```java
public void writeClass(MemoryBuffer buffer, ClassInfo classInfo) {
  if (classInfo.classId == NO_CLASS_ID) { // no class id provided.
    // use classname
    if (metaContextShareEnabled) {
      buffer.writeByte(USE_CLASS_VALUE_FLAG);
      // FIXME(chaokunyang) Register class but not register serializer can't be used with
      //  meta share mode, because no class def are sent to peer.
      writeClassWithMetaShare(buffer, classInfo);
    } else {
      // if it's null, it's a bug.
      assert classInfo.packageNameBytes != null;
      metaStringResolver.writeMetaStringBytesWithFlag(buffer, classInfo.packageNameBytes);
      assert classInfo.classNameBytes != null;
      metaStringResolver.writeMetaStringBytes(buffer, classInfo.classNameBytes);
    }
  } else {
    // use classId
    buffer.writeVarUint32(classInfo.classId << 1);
  }
}
```

**Comment by LiangliangSui** [comment]
> We can check meta size before writing meta

Done

File: java/fury-core/src/main/java/org/apache/fury/type/Constants.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.fury.type;
+
+public class Constants {
+  public static final int INVALID_OFFSITE = -1;
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/type/Constants.java:23]: I perfer not expose this variable. It's an internal implementation, we'd better keep it private

**Comment by LiangliangSui** [comment]
Hi @chaokunyang , could you please take a look?

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 4]
```diff
@@ -339,12 +341,14 @@ private void write(MemoryBuffer buffer, Object obj) {
...
-    if (!refResolver.writeRefOrNull(buffer, obj)) {
+    boolean isNull = refResolver.writeRefOrNull(buffer, obj);
+    if (!isNull) {
       ClassInfo classInfo = classResolver.getOrUpdateClassInfo(obj.getClass());
       classResolver.writeClass(buffer, classInfo);
       writeData(buffer, classInfo, obj);
     }
-    if (shareMeta) {
+    MetaContext metaContext = getSerializationContext().getMetaContext();
+    if (shareMeta && !isNull && metaContext != null && !metaContext.writingClassDefs.isEmpty()) {
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:351]:
```suggestion
    if (shareMeta && !metaContext.writingClassDefs.isEmpty()) {
```

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 5]
```diff
@@ -339,12 +341,14 @@ private void write(MemoryBuffer buffer, Object obj) {
       buffer.writeInt32(-1); // preserve 4-byte for nativeObjects start offsets.
     }
     // reduce caller stack
-    if (!refResolver.writeRefOrNull(buffer, obj)) {
+    boolean isNull = refResolver.writeRefOrNull(buffer, obj);
+    if (!isNull) {
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:345]:
```suggestion
    if (!refResolver.writeRefOrNull(buffer, obj)) {
```

File: java/fury-core/src/main/java/org/apache/fury/Fury.java (hunk 4 above)
Comment (LiangliangSui) [reply java/fury-core/src/main/java/org/apache/fury/Fury.java:351]: When obj is null, can we also not execute `classResolver.writeClassDefs(buffer)`?

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/Fury.java:351]: When obj is null, the writingClassDefs is empty

Comment (LiangliangSui) [reply java/fury-core/src/main/java/org/apache/fury/Fury.java:351]: got, `metaContext.writingClassDefs` may cause NPE risk, maybe we should check `assert metaContext != null` before using `metaContext.writingClassDefs`

Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/Fury.java:351]: then we can just write `metaContext != null && !metaContext.writingClassDefs.isEmpty()`

Comment (LiangliangSui) [reply java/fury-core/src/main/java/org/apache/fury/Fury.java]: updated

**Comment by LiangliangSui** [comment]
The above comments has been completed, please help review again.

**Comment by LiangliangSui** [comment]
@chaokunyang please help review this

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 6]
```diff
@@ -344,7 +346,8 @@ private void write(MemoryBuffer buffer, Object obj) {
       classResolver.writeClass(buffer, classInfo);
       writeData(buffer, classInfo, obj);
     }
-    if (shareMeta) {
+    MetaContext metaContext = getSerializationContext().getMetaContext();
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:349]:
```suggestion
    MetaContext metaContext = serializationContext.getMetaContext();
```

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 7]
```diff
@@ -1226,7 +1232,7 @@ public Object deserializeJavaObjectAndClass(MemoryBuffer buffer) {
     } catch (Throwable t) {
       throw ExceptionUtils.handleReadFailed(this, t);
     } finally {
-      if (shareMeta) {
+      if (shareMeta && classDefEndOffset != -1) {
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:1235]:
```suggestion
      if (classDefEndOffset != -1) {
```

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 8]
```diff
@@ -1116,7 +1122,7 @@ public <T> T deserializeJavaObject(MemoryBuffer buffer, Class<T> cls) {
...
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:1125]:
```suggestion
      if (classDefEndOffset != -1) {
```

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 9]
```diff
@@ -1063,9 +1066,12 @@ public void serializeJavaObject(MemoryBuffer buffer, Object obj) {
         if (!refResolver.writeRefOrNull(buffer, obj)) {
           ClassInfo classInfo = classResolver.getOrUpdateClassInfo(obj.getClass());
           writeData(buffer, classInfo, obj);
+          MetaContext metaContext = getSerializationContext().getMetaContext();
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:1069]:
```suggestion
          MetaContext metaContext = serializationContext.getMetaContext();
```

File: java/fury-core/src/main/java/org/apache/fury/Fury.java [hunk 10]
```diff
@@ -792,7 +795,7 @@ public Object deserialize(MemoryBuffer buffer, Iterable<MemoryBuffer> outOfBandB
     } catch (Throwable t) {
       throw ExceptionUtils.handleReadFailed(this, t);
     } finally {
-      if (shareMeta) {
+      if (shareMeta && classDefEndOffset != -1) {
```
Comment (chaokunyang) [inline java/fury-core/src/main/java/org/apache/fury/Fury.java:798]:
```suggestion
      if (classDefEndOffset != -1) {
```

**Review by chaokunyang — APPROVED** [review]
LGTM

**Pull request merged**